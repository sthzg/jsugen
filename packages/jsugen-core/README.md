# `jsugen-core`

> This package contains logic reused in various sub projects.
