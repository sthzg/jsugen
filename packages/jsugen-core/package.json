{
  "name": "@sthzg/jsugen-core",
  "version": "0.17.1",
  "description": "> TODO: description",
  "author": "Stephan Herzog <sthzgvie@gmail.com>",
  "homepage": "https://github.com/sthzg/jsugen#readme",
  "license": "MIT",
  "main": "lib/index.js",
  "publishConfig": {
    "access": "public"
  },
  "directories": {
    "lib": "lib"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/sthzg/jsugen.git"
  },
  "scripts": {
    "test": "echo \"Error: run tests from root\" && exit 1"
  },
  "bugs": {
    "url": "https://github.com/sthzg/jsugen/issues"
  },
  "dependencies": {
    "@babel/core": "^7.18.6",
    "@babel/plugin-proposal-class-properties": "^7.18.6",
    "@babel/plugin-proposal-export-namespace-from": "^7.18.6",
    "@babel/plugin-syntax-dynamic-import": "^7.8.3",
    "@babel/plugin-syntax-export-default-from": "^7.18.6",
    "@babel/plugin-syntax-import-meta": "^7.10.4",
    "@babel/plugin-transform-modules-commonjs": "^7.18.6",
    "@babel/preset-env": "^7.18.6",
    "@cloudflare/json-schema-walker": "^0.1.1",
    "babel-loader": "^8.2.5",
    "es6-error": "^4.1.1",
    "fs-extra": "^8.1.0",
    "glob": "^8.0.3",
    "json-schema-ref-parser": "^7.1.1",
    "json5-loader": "^3.0.0",
    "lodash": "^4.17.21",
    "prettier": "^1.17.0",
    "require-from-string": "^2.0.2",
    "rxjs": "^6.5.3",
    "webpack": "^5.73.0"
  }
}
