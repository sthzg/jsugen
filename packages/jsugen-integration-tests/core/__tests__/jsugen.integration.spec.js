import fs from 'fs-extra';
import { generate, ModuleFormat } from '@sthzg/jsugen-core';
import path from 'path';
import { byOutDirectory, withSnapshotCheck } from '../../testUtils/helpers';

/**
 * Set to `true` to to keep the generated files in `OUT_DIR`.
 */
const SKIP_DELETE = false;

const OUT_DIR = byOutDirectory('core');
const byPathInOut = needle => path.resolve(__dirname, OUT_DIR, needle);
const existsInOut = needle => fs.existsSync(byPathInOut(needle));

afterEach(() => {
  if (!SKIP_DELETE) {
    fs.emptyDirSync(OUT_DIR);
  }
});

// ---
// Dry Run Tests.
// ---
describe('main: dry run', () => {
  const config = {
    dryRun: true,
    silent: true,
    definitions: [
      {
        files: ['**/globTesting/**/*.schema.json'],
        generators: [
          '@sthzg/jsugen-generate-selector-functions',
          '@sthzg/jsugen-generate-lodash-object-paths',
        ],
        output: {
          baseDirectory: OUT_DIR,
        },
      },
    ],
  };

  test('should glob the expected files and run the configured generators', done => {
    expect.assertions(1);

    generate({ config }).subscribe(withSnapshotCheck(done, 'any'));
  });

  test('should respect the moduleFormat: cjs setting', done => {
    expect.assertions(1);

    generate({
      config: { ...config, moduleFormat: ModuleFormat.CJS },
    }).subscribe(withSnapshotCheck(done, 'expect the code to be transpiled'));
  });

  test('should respect the config.imports.lodashGet setting', done => {
    expect.assertions(1);

    generate({
      config: {
        ...config,
        imports: { lodashGet: `import { get } from 'lodash';` },
      },
    }).subscribe(
      withSnapshotCheck(done, 'expect the customized import statement'),
    );
  });
});

// ---
// IO Tests: Static output directory configuration.
// ---
describe('main: static output directory', () => {
  const config = {
    dryRun: false,
    silent: true,
    definitions: [
      {
        files: ['**/globTesting/**/*.schema.json'],
        generators: ['@sthzg/jsugen-generate-selector-functions'],
        output: {
          baseDirectory: OUT_DIR,
        },
      },
    ],
  };

  test('should generate the files at the expected location for output mode: static', done => {
    expect.assertions(2);

    generate({ config }).subscribe({
      complete() {
        expect(existsInOut('Hello/selectorFunctions.js')).toBe(true);
        expect(existsInOut('World/selectorFunctions.js')).toBe(true);

        done();
      },
    });
  });
});

// ---
// IO Tests: Source-file relative output directory configuration.
// ---
describe('main: colocated writes', () => {
  const config = {
    dryRun: false,
    silent: true,
    definitions: [
      {
        files: ['**/colocatedOutput/**/*.schema.json'],
        generators: ['@sthzg/jsugen-generate-selector-functions'],
        output: {
          baseDirectory:
            '<sourceFile>../../../out/core/relative-to-source-file',
        },
      },
    ],
  };

  test('should generate the files at the expected location for output mode: relative', done => {
    expect.assertions(2);

    generate({ config }).subscribe({
      complete() {
        expect(
          existsInOut('relative-to-source-file/Hello/selectorFunctions.js'),
        ).toBe(true);
        expect(
          existsInOut('relative-to-source-file/World/selectorFunctions.js'),
        ).toBe(true);

        done();
      },
    });
  });
});
