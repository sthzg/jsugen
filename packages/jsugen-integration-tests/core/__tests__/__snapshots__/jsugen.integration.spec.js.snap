// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main: dry run should glob the expected files and run the configured generators: any 1`] = `
"/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

import get from 'lodash-es/get';

export function byHello(property, defaultReturn) {
  return get(property, ['hello'], defaultReturn);
}

/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

export const byHelloPath = () => ['hello'];

/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

import get from 'lodash-es/get';

export function byWorld(property, defaultReturn) {
  return get(property, ['world'], defaultReturn);
}

/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

export const byWorldPath = () => ['world'];
"
`;

exports[`main: dry run should respect the config.imports.lodashGet setting: expect the customized import statement 1`] = `
"/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

import { get } from 'lodash';

export function byHello(property, defaultReturn) {
  return get(property, ['hello'], defaultReturn);
}

/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

export const byHelloPath = () => ['hello'];

/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

import { get } from 'lodash';

export function byWorld(property, defaultReturn) {
  return get(property, ['world'], defaultReturn);
}

/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

export const byWorldPath = () => ['world'];
"
`;

exports[`main: dry run should respect the moduleFormat: cjs setting: expect the code to be transpiled 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true,
});
exports.byHello = byHello;

var _get = _interopRequireDefault(require('lodash-es/get'));

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj };
}

/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */
function byHello(property, defaultReturn) {
  return (0, _get.default)(property, ['hello'], defaultReturn);
}

'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true,
});
exports.byHelloPath = void 0;

/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */
const byHelloPath = () => ['hello'];

exports.byHelloPath = byHelloPath;

'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true,
});
exports.byWorld = byWorld;

var _get = _interopRequireDefault(require('lodash-es/get'));

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj };
}

/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */
function byWorld(property, defaultReturn) {
  return (0, _get.default)(property, ['world'], defaultReturn);
}

'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true,
});
exports.byWorldPath = void 0;

/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */
const byWorldPath = () => ['world'];

exports.byWorldPath = byWorldPath;
"
`;
